[JIT] polish up overloads on free functions

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30356 [JIT] polish up overloads on free functions**
* #30749 add tests that schemas are valid

This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions.

Differential Revision: [D18864774](https://our.internmc.facebook.com/intern/diff/D18864774)

Commit:
polish up _overload on free functions

[ghstack-poisoned]

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

driazati reviewed (commented):
I can't find any info on how default args should be handled in either the mypy docs or PEP 484, but why do the defaults need to be on the implementation? Doesn't that prevent something like the following?

```python
@overload
def fn(x: int = 2): pass

@overload
def fn(x: str = "hi"): pass

def fn(x):
	...
```

eellison commented:
> I can't find any info on how default args should be handled in either the mypy docs or PEP 484, but why do the defaults need to be on the implementation? Doesn't that prevent something like the following?
> 
> ```python
> @overload
> def fn(x: int = 2): pass
> 
> @overload
> def fn(x: str = "hi"): pass
> 
> def fn(x):
> 	...
> ```

The overload declarations get ignored at runtime, so it's not valid to put default arguments on them

eellison commented:
> > I can't find any info on how default args should be handled in either the mypy docs or PEP 484, but why do the defaults need to be on the implementation? Doesn't that prevent something like the following?
> > ```python
> > @overload
> > def fn(x: int = 2): pass
> > 
> > @overload
> > def fn(x: str = "hi"): pass
> > 
> > def fn(x):
> > 	...
> > ```
> 
The overload declarations get ignored at runtime, so it's not valid to put default arguments on them. In your example there's no invocation of the function in which the default arguments have any effect.

eellison commented on test/test_jit.py:
```diff
@@ -14869,16 +14869,28 @@ def invoke_function():
+        #
         @torch.jit._overload  # noqa: F811
-        def identity(x1=1.0):  # noqa: F811
+        def identity(x1):  # noqa: F811
```
bc i updated the behavior so that it errors when default arguments are put on overload declarations, which is the correct behavior. If JIT were to take them then there would be a divergence in behavior between pytorch eager & JIT, since the overloads have no runtime effect

eellison commented on test/test_jit.py:
```diff
@@ -14856,7 +14856,7 @@ def test_simple(x1):  # noqa: F811
 
         def test_simple(x1):  # noqa: F811
-            return x1 + 5
+            return x1
```
so that i could reuse the test with str.

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -145,7 +145,9 @@ struct FunctionSchema {
         is_vararg_(is_vararg),
-        is_varret_(is_varret) {}
+        is_varret_(is_varret) {
+    checkDefaultArgs();
```
it would be nice if i could put it in the compiler but we do the following pattern currently:

```
  auto defined_functions = cu->define(
      QualifiedName(name.prefix()),
      {def},
      {pythonResolver(std::move(rcb))},
      nullptr,
      true);
  TORCH_INTERNAL_ASSERT(defined_functions.size() == 1);
  auto& defined = defined_functions[0];
  defined->setSchema(getSchemaWithNameAndDefaults(
      def.range(), defined->getSchema(), def.name().name(), defaults));
```
default arguments are added to the schema following compilation, the compiler doesn't have access to it

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

suo commented:
I would say we should at least include free functions + module support in this PR. We can leave classes off for when we do a pass to make them better in general. I can review this PR as is, or would you prefer me wait until the module stuff is in as well?

eellison commented:
> I would say we should at least include free functions + module support in this PR. We can leave classes off for when we do a pass to make them better in general. I can review this PR as is, or would you prefer me wait until the module stuff is in as well?

There are a couple reasons to land this is as is without the equivalent changes to modules:

* I would like to use the api `@overload` for both modules and free-functions (no one liked `@script_method`). I forget the full details but I remember that I had I couldn't do this in python 2 because it's difficult to determine what class a method is defined in at the point it's defined. If we're waiting until we drop python 2 that means we won't be able to release this api for a month. 

* The work to add this to UDT might end up refactoring out the work to add this to modules (this could also not be true). 

There are people internally who've asked for overloads just on free functions. I think it would be fine to land this as is and then not fully document it until it's also supported for modules. At any rate yes I would like a review for this as a stand alone PR.

suo reviewed (changes requested):
Overall change looks reasonable, and I think it's great that we're being careful about conforming to what people expect. I don't understand the full compilation flow though—left some comments in line asking questions.

eellison commented on torch/jit/__init__.py:
```diff
@@ -2034,36 +2035,42 @@ def _check_no_signature(func):
     compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
-    if compiled_overloads is not None:
+    uncompiled_overloads = _jit_internal._get_fn_overloads(qual_name)
+    if compiled_overloads is not None and uncompiled_overloads is None:
```
I don't think this is a real use case, but you suggested it as part of the previous PR. https://github.com/pytorch/pytorch/pull/23886#discussion_r312097222

Agreed it would be good to build upon caching logic from other PR.

suo replied:
lol. Is this actually allowed to happen (dynamic registration of overloads) or was I being paranoid

eellison commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -266,6 +296,101 @@ FunctionSchema getSchemaWithNameAndDefaults(
+      overload_decl.return_type());
+}
+
+void checkSchema(const FunctionSchema& schema, const SourceRange& range) {
```
I tried this in Function Schema originally, but i failed on a bunch of functions for reasons I couldn't really understand. I can debug this more...

eellison replied:
It fails because symbolic_script registers an invalid schema:

```def __interpolate_0(input,
                            size: Optional[int],
                            scale_factor: Optional[List[float]],
                            mode: str='nearest',
                            align_corners: Optional[bool]):```

suo replied:
ah, can we fold that fix into here? Seems bad to have it out there lol

eellison commented on torch/jit/__init__.py:
```diff
@@ -2034,36 +2035,42 @@ def _check_no_signature(func):
         raise RuntimeError("Must explicitly add type annotations to overloaded functions: {}".format(qual_name))
 
-def _get_overload_decl_and_defaults(func):
+def _check_no_default_args(func):
```
There are many errors we throw in python, and this is a very simple function. I don't think we need to move it all to C++. Agreed source range should be a part of the error here though, will update it to use `FrontendError` so it does.

suo replied:
It's not so much about the the complexity, it would just be nice to put error checking in one place so future readers don't need to figure out whether it was already done in the other place.

eellison commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -266,6 +296,101 @@ FunctionSchema getSchemaWithNameAndDefaults(
       schema.is_varret());
 }
 
+static Decl computeOverloadDecl(
```
it's not just merging default arguments, it's also merging extra parameters from the implementation def

eellison replied:
the alternative is `mergeDefaultsAndExtraParametersIntoOverload` , which is a little too long imo...

suo replied:
In general I think it's better to have a long name than a non-descriptive one. Because if it's non-descriptive, then the user has to read the entire function to find out what it does, instead of just having to parse a long name.

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

eellison reviewed (commented):
@suo i addressed msot of your comments, take another look ?

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -266,6 +296,101 @@ FunctionSchema getSchemaWithNameAndDefaults(
+  }
+}
+
+static StrongFunctionPtr script_compile_overloaded_function(
```
I think we should do that. I previously thought of Python resolver as "part of the binding code" but if it's causing us to put a bunch of extra logic in init.cpp it deserves to be factored out. I think it's fine to do this in a follow-up though.

suo reviewed (approved):
lgtm!

Commit:
Update on "[JIT] polish up overloads on free functions"


This finishes up the `torch.jit.overload` api for free-functions.
- defaults now required on the implementation function itself
- fully follows [overload spec](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) such that the following is supported 

```
@overload
def mouse_event(x1: int, y1: int) -> ClickEvent: ...
def mouse_event(x1: int,
                y1: int,
                x2: Optional[int] = None,
                y2: Optional[int] = None): ...
```

Note: `jit.overload` isn't supported yet for UDT, but is support for modules. This PR doesn't make the same changes for modules, if reviewers think I should include them then I could do so in a follow up PR or wait to land this. Since that's still an internal api I think it's fine, and the changes here would allow us to expose `torch.jit.overload` on free functions. 



[ghstack-poisoned]

wanchaol commented:
I am reverting this PR bc it broke the master. Can you rebase and try land again?
